<a href="https://colab.research.google.com/github/kkidane002/summer_6211/blob/main/KK_dsba6211_summer2024_lab2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lab 2 Instructions

For these problems, you'll use [scikit-learn](https://scikit-learn.org/stable/). Feel free to use the documentation as reference.

There are two parts to do:
1. code fill-in the blank
2. markdown written comments

Code fill-in the blanks are code snippets with:
```
# fill in the blank:
```

For markdown written comments, replace the text with:

`<!-- Write in markdown your answer below -->`

As normal, please complete this Colab notebook, save it to your DSBA6211 GitHub repo, and then submit that URL to that notebook on Canvas.


# Problem 1

A pharmaceutical company is conducting a market research project to gauge consumer interest in buying a **cold sore healthcare product**. To do so, they've collected data from a sample of **1,023 consumers**.

As part of the study, each participant was asked to evaluate **35 key factors** that influence their decision to purchase a cold sore treatment, rating each factor on a **10-point scale** to indicate its level of importance.

Additionally, they were asked to rate their overall interest in buying the product (y). This is the variable we're most interested in analyzing.

In [3]:
import pandas as pd

data_url = "https://github.com/DS-100/textbook/raw/master/content/datasets/market-analysis.csv"

df = pd.read_csv(data_url)

df.head(n=5)

y  x1  x2  x3  x4  x5  x6  x7  x8  x9  ...  x26  x27  x28  x29  x30  x31  \
0  7   8   8   8   8   8   8   8   8   8  ...    8    8    8    7    8    8   
1  9   9   9   9   9   9   7   9   9   9  ...    9    9    9    9    9    9   
2  5   6   6   5   5   8   5   7   5   3  ...    8    8    6    7    7    5   
3  6   7   5   8   6   8   6   7   4   6  ...    7    7    6    6    7    8   
4  5   4   6   4   4   3   4   5   4   4  ...    5    5    5    5    5    4   

   x32  x33  x34  x35  
0    8    5    8    8  
1    9    9    3    9  
2    8    5    4    4  
3    6    3    6    8  
4    6    4    5    6  

[5 rows x 36 columns]

Here's a table that provides a description for each of the 35 variables and its correlation with `y`.

|  | Corr | Description |  | Corr | Description |
| --- | --- | --------- | --- | --- | --------- |
| x1  | 0.70 | provides soothing relief | x19 | 0.54 | has a non-messy application |
| x2  | 0.58 | moisturizes cold sore blister | x20 | 0.70 | good for any stage of a cold |
| x3  | 0.69 | provides long-lasting relief | x21 | 0.49 | easy to apply/take |
| x4  | 0.70 | provides fast-acting relief | x22 | 0.52 | package keeps from contamination |
| x5 | 0.72 | shortens duration of a cold | x23 | 0.57 | easy to dispense a right amount |
| x6  | 0.68 | stops the virus from spreading | x24 | 0.63 | worth the price it costs |
| x7 | 0.67| dries up cold sore | x25 | 0.57 | recommended most by pharamacists |
| x8 | 0.72 | heals fast | x26 | 0.54 | recommended by doctors |
| x9 | 0.72 | penetrates deep | x27 | 0.54 | FDA approved |
| x10 | 0.65 | relieves pain | x28 | 0.64 | a brand I trust |
| x11 |0.61 | prevents cold | x29 | 0.60 | clinically proven |
| x12 | 0.73 | prevents from getting worse | x30 | 0.68 | a brand I would recommend |
| x13 | 0.57 | medicated | x31 | 0.74 | an effective treatment |
| x14 | 0.61 | prescription strength | x32  |0.37 | portable |
| x15 | 0.63 | repairs damaged skin | x33 | 0.37 | discreet packaging |
| x16 | 0.67 | blocks virus from spreading | x34 | 0.55 | helps conceal cold sores |
| x17 | 0.42 | contains SPF | x35 | 0.63 | absorbs quickly |
| x18 | 0.57 | non-irritating | | | |

You can replicate the correlations like this:

In [4]:
round(df[['x1','x2','y']].corr(),2)

x1    x2     y
x1  1.00  0.69  0.70
x2  0.69  1.00  0.58
y   0.70  0.58  1.00

For this question, we'll convert the y variable to a binary to help us frame the problem as a binary classification.

In [5]:
# if df["y"] > 7 then 1
y = df["y"].apply(lambda x: 1 if x > 7 else 0)
X = df.drop(columns=["y"])

## Fill in the blanks

Create a `train_test_split` on your `X` and `y` data. Set the `test_size=200` and set the `random_state` to `42`.

In [7]:
# fill in blank:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=200, random_state=42)


Create scikit-learn pipeline with StandardScalar and LogisticRegression. Use 42 for the `random_state`.



In [12]:
# fill in blank:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression


pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('logreg', LogisticRegression(random_state=42))
])

pipeline.fit(X_train,y_train)

Pipeline(steps=[('scaler', StandardScaler()),
                ('logreg', LogisticRegression(random_state=42))])

Measure test (out-of-sample) accuracy using `accuracy_score`.

In [14]:
# fill in blank:
from sklearn.metrics import accuracy_score
# we can now use it like any other estimator
accuracy_score(pipeline.predict(X_test), y_test)


0.79

Let's now explore what happens if we modify regularization by modifying hyperparameters.

For this, I recommend referencing [this tutorial on scikit-learn hyperparameters](https://inria.github.io/scikit-learn-mooc/python_scripts/parameter_tuning_manual.html)

Show all of the parameters names (hint: you'll need to find the method on `pipe` that displays all of `pipe`'s parameters).

In [15]:
# fill in the blank

pipeline_parameters = pipeline.get_params()
for param_name in pipeline_parameters:
    print(param_name)

memory
steps
verbose
scaler
logreg
scaler__copy
scaler__with_mean
scaler__with_std
logreg__C
logreg__class_weight
logreg__dual
logreg__fit_intercept
logreg__intercept_scaling
logreg__l1_ratio
logreg__max_iter
logreg__multi_class
logreg__n_jobs
logreg__penalty
logreg__random_state
logreg__solver
logreg__tol
logreg__verbose
logreg__warm_start


With the name of the C hyperparameter from the previous step, use the `set_params` to train a model with a C value of `100`.

In [16]:
# fill in the blank:
pipeline.set_params(logreg__C=100)
pipeline.fit(X_train, y_train)



Pipeline(steps=[('scaler', StandardScaler()),
                ('logreg', LogisticRegression(C=100, random_state=42))])

Now use `cross_validate` (I've provided the import below) to do cross-validation to test these different values of C: `[1e-3, 1e-2, 1e-1, 1, 10, 100]`

You will need to provide the mean and +/- 1 std dev for each value of C. Be sure to carefully select the right datasets to use.

(Hint: see the tutorial on scikit-learn hyperparameters provided above)

In [18]:
from sklearn.model_selection import cross_validate
import numpy as np

# fill in the blanks:

results = {}

C_values = [1e-3, 1e-2, 1e-1, 1, 10, 100]
for C in C_values:
    pipeline.set_params(logreg__C=C)
    cv_results = cross_validate(pipeline, X_train, y_train, cv=5, return_train_score=False)
    results[C] = cv_results

for C, cv_results in results.items():
    mean_score = np.mean(cv_results['test_score'])
    std_score = np.std(cv_results['test_score'])
    print(f"C={C}: Mean CV Score = {mean_score:.4f} +/- {std_score:.4f}")

C=0.001: Mean CV Score = 0.8178 +/- 0.0256
C=0.01: Mean CV Score = 0.8226 +/- 0.0207
C=0.1: Mean CV Score = 0.8117 +/- 0.0252
C=1: Mean CV Score = 0.8044 +/- 0.0216
C=10: Mean CV Score = 0.8032 +/- 0.0232
C=100: Mean CV Score = 0.8032 +/- 0.0232


Based on your cross-validation, which hyperparameter would you suggest to use? How large of an impact do the different possible C values have?

`<!-- Write in markdown your answer below -->`

Based on the cross-validation results, the hyperparameter value of **C=0.01** is suggested to use, as it has the highest mean CV score of **0.8226** with a standard deviation of 0.0207.

The impact of different C values on the model performance is noticeable but not extremely large. The mean CV scores range from **0.8032 to 0.8226**, with a difference of about 0.02. This indicates that while tuning C does have an effect on model performance, the overall variation in accuracy is relatively moderate across the different values tested.

# Problem 2: problems in pipelines

Let's assume that you have run this model.

In [19]:
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split

random_state = 42
X, y = make_classification(random_state=random_state, n_features=10)
print(X.shape, y.shape)

(100, 10) (100,)


In [20]:
print(X[0:5],y[0:5])

[[-1.14052601  1.35970566  0.86199147  0.84609208  0.60600995 -1.55662917
   1.75479418  1.69645637 -1.28042935 -2.08192941]
 [-0.07873421 -1.32933233  0.6273745  -1.19300559 -0.77300978  0.0976761
   0.49799829  0.95927083  0.02451017  1.45114361]
 [ 0.80742726  0.73019848 -1.28568005  0.88948365 -1.8048821  -0.76325916
   0.04808495 -0.90431663 -1.62754244  0.2597225 ]
 [ 0.58846525 -0.3751207  -0.57500215 -0.14951801 -0.56372455  0.41293145
   0.24368721 -0.50694318 -0.8222204   0.24496657]
 [ 1.63631177 -1.64060704 -1.36045573 -0.94116325 -1.43014138  1.6324113
   0.13074058 -1.43586215 -0.44004449  1.44127329]] [1 1 0 0 0]


We'll partition the data into a train-test split with 30% in the test dataset.

In [21]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=random_state)

Let's now assume that you were told to standardize your X variable like so:

In [22]:
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_transformed = scaler.fit_transform(X_train)
model = LogisticRegression(C=1).fit(X_train_transformed, y_train)
accuracy_score(y_test, model.predict(X_test))

0.9666666666666667

## Question

What is a problem with the above workflow?

*Hint: go step by step, and view all of the outputs and inputs*

`<!-- Write in markdown your answer below -->`

Write in the code snippet below a corrected version of your model.

*Hint: rewrite to fix the problem (2 lines of code)*

The primary problem with the above workflow is that the **StandardScaler** is fitted only on the training data and then used to transform the training data, but the test data (**X_test**) is not scaled using the same **StandardScaler** instance. This means that the scaling applied to the training data is not consistently applied to the test data, which can lead to incorrect model evaluation.

In [23]:
# fill in the blank:
X_test_transformed = scaler.transform(X_test)
accuracy_score(y_test, model.predict(X_test_transformed))

0.9333333333333333

# Problem 3: Building pipelines

For this problem, we'll use `skrub`, a helpful `scikit-learn` extension package.


In [24]:
!pip install skrub

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 142.6/142.6 kB 2.9 MB/s eta 0:00:00


In [25]:
from sklearn.datasets import fetch_openml

X, y = fetch_openml("titanic", version=1, as_frame=True, return_X_y=True)

X.head(3)

/usr/local/lib/python3.10/dist-packages/sklearn/datasets/_openml.py:968: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


pclass                            name     sex      age  sibsp  parch  \
0     1.0   Allen, Miss. Elisabeth Walton  female  29.0000    0.0    0.0   
1     1.0  Allison, Master. Hudson Trevor    male   0.9167    1.0    2.0   
2     1.0    Allison, Miss. Helen Loraine  female   2.0000    1.0    2.0   

   ticket      fare    cabin embarked  boat  body  \
0   24160  211.3375       B5        S     2  None   
1  113781  151.5500  C22 C26        S    11   NaN   
2  113781  151.5500  C22 C26        S  None   NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON

Create a train vs test partition, holding out 20% of the data.

In [27]:
# fill in the blank:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Use the code below to create a simple `scikit-learn` pipeline with only `pclass` and `sex` encoded as a One Hot Encoding:

In [29]:
# fill in the blank:

from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder

# Define the column transformer to apply One-Hot Encoding to `pclass` and `sex`
preprocessor = ColumnTransformer(
    transformers=[
        ('ohe', OneHotEncoder(), ['pclass', 'sex'])
    ]
)

pipe = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', LogisticRegression(random_state=42))
])

Next, run `.fit(X_train)` on your pipe.

In [30]:
# fill in the blank:

pipe.fit(X_train,y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('ohe', OneHotEncoder(),
                                                  ['pclass', 'sex'])])),
                ('classifier', LogisticRegression(random_state=42))])

Now combine your `pclass`-`sex` pipeline with a union of another pipeline, this time with `age` and `fare` that are normalized with `StandardScalar`:

In [31]:
# fill in the blank:

X = X[['pclass', 'sex', 'age', 'fare']]

# Handle missing values for age and fare
X['age'].fillna(X['age'].median(), inplace=True)
X['fare'].fillna(X['fare'].median(), inplace=True)

# Create a train-test split, holding out 20% of the data for testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Define the column transformer with two transformers: one for One-Hot Encoding and one for Standard Scaling
preprocessor = ColumnTransformer(
    transformers=[
        ('cat', OneHotEncoder(), ['pclass', 'sex']),
        ('num', StandardScaler(), ['age', 'fare'])
    ]
)

# Create a pipeline that includes the preprocessor and a classifier
pipe = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', LogisticRegression(random_state=42))
])

# Fit the pipeline on the training data
pipeline.fit(X_train, y_train)

<ipython-input-31-64eef34d7216>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X['age'].fillna(X['age'].median(), inplace=True)
<ipython-input-31-64eef34d7216>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X['fare'].fillna(X['fare'].median(), inplace=True)


Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('ohe', OneHotEncoder(),
                                                  ['pclass', 'sex'])])),
                ('classifier', LogisticRegression(random_state=42))])

Now change `pipe` so that there is a `HistGradientBoostingClassifier` at the end of the pipeline. Then fit your model using

In [ ]:
from sklearn.ensemble import HistGradientBoostingClassifier

# fill in the blank:

pipe = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', HistGradientBoostingClassifier(random_state=42))
])

Now `.fit()` your pipeline based on your training data. Define that object as `model`.

In [32]:
# fill in the blank

model = pipe.fit(X_train, y_train)

Use your `X_test` data to make prediction on the test data and name it `y_pred`.

Then use `scikit-learn`'s `classification_report` (import provided) to view the model's performance. Be sure to run it as `print(classification_report ... )` so that it prints to the console.

In [33]:
from sklearn.metrics import classification_report

# fill in the blank
y_pred = pipe.predict(X_test)
print(classification_report(y_test, y_pred))


              precision    recall  f1-score   support

           0       0.75      0.89      0.81       144
           1       0.82      0.64      0.72       118

    accuracy                           0.77       262
   macro avg       0.79      0.76      0.77       262
weighted avg       0.78      0.77      0.77       262



Based on your results, does your model have higher precision or recall when Survival = 1?

`<!-- Write in markdown your answer below -->`

In this case, when Survival = 1, the precision **(0.82)** is higher than the recall (**0.64)**, indicating that the model is better at correctly identifying instances of survival (true positives) among all predicted survival instances than it is at capturing all actual survival instances in the dataset.

Now take the classification report output and use whatever Large Language Model (ChatGPT, Claude, LLaMa3, whatever). Generate a custom prompt where you provide the LLM the definition of the Titantic dataset and what 1 (Survival) and 0 (Non-Survival) mean. Ask the LLM to now interpret your model's results.

`<!-- Show your Prompt / LLM below -->`

`<!-- Show your the LLM's response -->`

Do you agree with the LLM's interpretation?

`<!-- Write in markdown your answer below -->`

**PROMPT**

Based on the result above, interpret the model? iT is a result of a titanic dataset where in the predictive class 1 indicates survival and 0 is not survived?

**LLM Response**

Based on the provided classification report for the Titanic dataset, we can interpret the model's performance as follows:

Accuracy: The overall accuracy of the model on the test set is 0.77, indicating that approximately 77% of the predictions made by the model are correct.

Precision and Recall:

For the class representing passengers who did not survive (0):
Precision: 0.75
This means that out of all instances predicted as not survived by the model, 75% are actually correct.
Recall: 0.89
This means that out of all actual instances where passengers did not survive, the model correctly identifies 89% of them.
For the class representing passengers who survived (1):
Precision: 0.82
This means that out of all instances predicted as survived by the model, 82% are actually correct.
Recall: 0.64
This means that out of all actual instances where passengers survived, the model correctly identifies 64% of them.
F1-Score: The F1-score is the harmonic mean of precision and recall. It provides a balance between precision and recall. The weighted average F1-score for this model is 0.77.

Support: The number of instances (passengers) in each class in the test set is provided as support.

In summary, the model performs reasonably well in terms of accuracy, precision, recall, and F1-score. However, there is a trade-off between precision and recall, especially for the class representing passengers who survived. The model has higher precision for predicting survival, but it sacrifices some recall, meaning it may miss some actual instances of survival. Further analysis and tuning may be necessary to improve the model's performance.


**LLM interpretation response**

Overall, I agree with the LLM's interpretation of the result.

# Bonus Problem: Feature Selection

Assume you have the following data for a classification problem.

You've found [`SelectKBest`](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.SelectKBest.html) which will do feature selection based on some metric.

Let's assume you have a big dataset of 200 data points, 10000 features (X variables) and 2 levels in a y variable.



In [ ]:
import numpy as np

n_samples, n_features, n_classes = 200, 10000, 2
rng = np.random.RandomState(42)
X = rng.standard_normal((n_samples, n_features))
y = rng.choice(n_classes, n_samples)

print(X.shape, y.shape)

(200, 10000) (200,)


In [ ]:
print(X[0:10],y[0:10])

[[ 0.49671415 -0.1382643   0.64768854 ... -0.70531672  0.49576557
   0.64438845]
 [-0.67849473 -0.30549946 -0.59738106 ...  0.37835397  1.71352973
  -1.6199198 ]
 [ 0.34828625  0.28332359 -0.93651985 ... -0.50136732 -1.89235682
  -0.22888421]
 ...
 [ 0.32015241  1.1154616  -1.50523819 ... -0.46895701 -0.00334278
  -0.99348803]
 [ 0.66400108 -0.29611605 -0.06658641 ... -0.29638908  0.25959623
  -0.19642193]
 [ 0.96089457 -0.36996474 -0.57958089 ...  0.40918508 -0.21109167
   0.12006294]] [1 1 1 0 1 1 1 0 1 1]


In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.feature_selection import SelectKBest
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import accuracy_score

X_selected = SelectKBest(k=25).fit_transform(X, y)

X_train, X_test, y_train, y_test = train_test_split(
    X_selected, y, random_state=42)
gbc = GradientBoostingClassifier(random_state=42)
gbc.fit(X_train, y_train)

y_pred = gbc.predict(X_test)
accuracy_score(y_test, y_pred)

0.76

## Question

What is a problem with the workflow above?

`<!-- Write in markdown your answer below -->`

Rewrite the workflow to correct the workflow.

In [ ]:
# fill in the blank: